thecourseforum/theCourseForum2
# Save Course

## Status

Done

## GitHub Issues addressed

- This PR closes #265 

## What I did
Added Save Course Feature 
- On course-instructor page user can click save course (requires login)
- Optional notes field can be saved with that course-instructor pair
- Under profile -> saved courses, user can view their saved courses
- On saved courses page, user can rearrange courses, sort courses, and edit notes

## Screenshots
Save Course Button
![image](https://user-images.githubusercontent.com/34071587/116001850-bd1e8880-a5c4-11eb-90e1-06118a3c311b.png)
![image](https://user-images.githubusercontent.com/34071587/116001865-cd366800-a5c4-11eb-8135-3f6944a30462.png)

Save Course Page
![image](https://user-images.githubusercontent.com/34071587/116001919-0a025f00-a5c5-11eb-9f7d-30f24af015bb.png)
![image](https://user-images.githubusercontent.com/34071587/116001904-f9ea7f80-a5c4-11eb-8aba-badc05f29a7b.png)

## Tests
- Models tested in `tcf_website/tests/test_saved_course.py`

**Frontend Tests**

Test Saving Courses
- Navigate to Saved Courses page and verify that you can access it/no courses are currently saved
- Navigate to any course-instructor page, verify that `Save Course` button exists
- Click button, add text to notes field, click save
- Button should change to `Unsave Course`
- Repeat this with as many courses as desired
- Navigate to Saved Courses page and verify that the saved course(s) and corresponding notes appear correctly

Testing Unsave
- Click `Unsave Course` on the course-instructor page, navigate to Saved Courses page, verify the saved course no longer appears
- Click `Unsave` on the Saved Courses page, verify the saved course no longer appears, refresh the page and verify again

Testing Editing Notes
- On Saved Courses pages, click on notes field, change notes, click save, verify that notes are updated, refresh, verify update is saved

Testing Sorting/Reordering (with more than one course saved)
- On Saved Courses page, click sorting buttons and verify courses appear in correct order
- Drag and drop course cards, refresh page, verfiy that ordering is saved. 

## Questions/Discussions/Notes
-

## Merging the PR

- Who should merge this PR?
- Should the commit history be preserved in the base branch, or is it okay to combine all of them into a single commit ("squash-merge")?
  - [] preserve the history

## Add your changes to "Next Update" on the [release history page](https://github.com/thecourseforum/theCourseForum2/wiki/Release-History) once your PR gets merged!
